### ISSUE TITLE
systemd-boot: no mechanism for disabling type1 or type2 enumeration

### ISSUE THREAD

**Issue by aubergino** [systemd/systemd#29799]
### systemd version the issue has been seen with

systemd-254.5-1-arch

### Used distribution

Arch Linux

### Linux kernel version used

6.5.9-zen2-1-zen

### CPU architectures issue was seen on

x86_64

### Component

systemd-boot

### Expected behaviour you didn't see

Some sort of explanatory note in the system journal (yes, I booted with systemd.log_level=debug), like 'Error parsing auto-entries config option', as suggested by https://github.com/systemd/systemd/blob/da79ae6fc5a785b8ef07ec9b736097bfba1805f0/src/boot/efi/boot.c#L1229C3-L1229C3 (please excuse the poor formatting). Systemd-boot seems to have understood my config file without issues.

### Unexpected behaviour you saw

- The two UKIs on my ESP are shown in the systemd-boot menu, even though I have one entries referring to each of them that only set the title and sort-key.
- The (theoretically) automatically detected Windows Boot Manager was not shown, presumably because it has an entry referring to it explicitly.

![auto-entries-fail](https://github.com/systemd/systemd/assets/113323010/f7b1313c-9d5f-4c65-94b7-4fdafc1ff25a)

### Steps to reproduce the problem

I use unsigned UKIs to boot Arch, and I have multiple systemd-boot entries used both to change the displayed title of each boot entry and to provide alternative cmdlines (via the 'options' section of each entry) for debugging and other purposes. All the entries use the same UKI, except the 'fallback' entry, which uses a UKI with the same kernel but different initrd and cmdline.
I have Windows and Debian installed in addition to Arch; Windows boots via the usual `/EFI/Microsoft/Boot/bootmgfw.efi` while Debian boots using a completely separate installation of GRUB.
Relevant config/outputs can be found attached.

1. Have my config, most saliently having 2 UKIs with entries explicitly referring to each of them with no additional options; and having the Windows Boot Manager installed in the normal place with an entry explicitly referring to it with no additional options
2. Match my loader.conf
3. Boot your computer

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
# ls -R /efi
.:
 EFI   loader  'System Volume Information'

./EFI:
BOOT  debian  Lenovo  Linux  Microsoft  systemd

./EFI/BOOT:
BOOTX64.EFI

./EFI/debian:
BOOTX64.CSV  fbx64.efi  grub.cfg  grubx64.efi  mmx64.efi  shimx64.efi

./EFI/Lenovo:
[[omitted for brevity]]

./EFI/Linux:
archlinux-linux-zen.efi  archlinux-linux-zen-fallback.efi

./EFI/Microsoft:
Boot  Recovery

./EFI/Microsoft/Boot:
bootmgfw.efi
bootmgr.efi 
[[rest of folder omitted for brevity]]

./EFI/systemd:
systemd-bootx64.efi

./loader:
entries  entries.srel  loader.conf  random-seed

./loader/entries:
arch.conf  arch-debug.conf  arch-fallback.conf  arch-multi-user.conf  debian.conf  windows.conf

'./System Volume Information':


# /loader/loader.conf
timeout         2
#console-mode   keep
default         arch.conf
editor          false
auto-entries    false
auto-firmware   false


# bootctl list
         type: Boot Loader Specification Type #1 (.conf)
        title: Arch Linux 10 (default)
           id: arch.conf
       source: /efi//loader/entries/arch.conf
     sort-key: 0arch
          efi: /efi//EFI/Linux/archlinux-linux-zen.efi

         type: Boot Loader Specification Type #1 (.conf)
        title: Windows 11
           id: windows.conf
       source: /efi//loader/entries/windows.conf
     sort-key: 1windows
          efi: /efi//EFI/Microsoft/Boot/bootmgfw.efi

[[omitted for brevity]]

         type: Boot Loader Specification Type #2 (.efi)
        title: Arch Linux (6.5.9-zen2-1-zen)
           id: archlinux-linux-zen.efi
       source: /efi//EFI/Linux/archlinux-linux-zen.efi
     sort-key: arch
      version: 6.5.9-zen2-1-zen
        linux: /efi//EFI/Linux/archlinux-linux-zen.efi
      options: rd.luks.name=[[omitted]]=dunsparce root=/dev/mapper/dunsparce rw quiet bgrt_disable rd.luks.options=timeout=120 ibt=off resume=UUID=[[omitted]] resume_offset=[[omitted]]

         type: Boot Loader Specification Type #2 (.efi)
        title: Arch Linux (6.5.9-zen2-1-zen~fallback)
           id: archlinux-linux-zen-fallback.efi
       source: /efi//EFI/Linux/archlinux-linux-zen-fallback.efi
     sort-key: arch
      version: 6.5.9-zen2-1-zen~fallback
        linux: /efi//EFI/Linux/archlinux-linux-zen-fallback.efi
      options: rd.luks.name=[[omitted]]=dunsparce root=/dev/mapper/dunsparce rw bgrt_disable systemd.journald.forward_to_console=1 console=tty1 pause_on_oops=5 loglevel=8 resume=UUID=[[omitted]] resume_offset=[[omitted]]
```

**Comment by poettering** [comment]
> * The two UKIs on my ESP are shown in the systemd-boot menu, even though I have one entries referring to each of them that only set the title and sort-key.

Hmm? type1 and type2 entries are both enumerated and added to the menu. What else did you expect? That we suppress type2 entries if a kernel in it is mentioned in a type1 entry? that makes little sense to me.

I think what might make sense to add is to suppress a type2 entry if an identically named type1 entry already exists. There's a TODO list item for that already.

> * The (theoretically) automatically detected Windows Boot Manager was not shown, presumably because it has an entry referring to it explicitly.

I am a bit confused? You set auto-entries to false, says the issue title? hence we don't search for a windows entry. i.e. all the "auto-*" entries are suppressed, hence you got what you wanted?

We could probably make that clearer in the docs, but to me everything appears to work the way it should?

**Comment by aubergino** [comment]
As far as I understood, type2 entries, the .efi ones, would be encompassed by the `auto-entries` configuration option, as they are  automatically detected by systemd-boot (rather than being explicitly specified in a type1 entry, which feels decidedly more manual and less automatic to me). As far as I _knew_, all .efi files were detected and enumerated in the same way, whether Windows Boot Manager, GRUB, or a UKI. So I suppose you're confused because I was confused.

I was trying to disable detection of type2 entries, so that I could specify the ones I wanted to appear as type1 entries (and part of that is setting titles that are a bit more pleasing to the eye). I agree that simply mentioning a kernel in a type1 entry shouldn't suppress a type2 entry, so the "TODO list item" (was it #24539 ?) you mentioned sounds good to me. To some extent, therefore, this issue is just a duplicate, and an RFE not a bug.

Maybe someone _should_ do something about the docs to clear up the fact that type2 entries are always automatically detected--I could have a look, but I wouldn't know where to go, could you give me a pointer please?

Thanks for the response!

**Comment by YHNdnzj** [comment]
> As far as I understood, type2 entries, the .efi ones, would be encompassed by the auto-entries configuration option, as they are automatically detected by systemd-boot (rather than being explicitly specified in a type1 entry, which feels decidedly more manual and less automatic to me).

No, you are confusing two fundamentally different things. UKIs, just like type1 entries, carry metadata. They're self-contained though, meaning that the metadata is in the image itself. Putting the UKI in the BLS-defined location _is_ explicitly specifying it.

**Comment by TriMoon** [comment]
I wasn't aware that one could change the options used for a UKI, isn't the purpose of UKI's to create a fixed image with fixed kernel-command-line options? :thinking:

**Comment by aubergino** [comment]
@YHNdnzj thank you for the incredibly helpful response

@TriMoon Apparently I misunderstood the purpose of a UKI, if that's it. It's just that my system root is on an encrypted drive using a format that GRUB doesn't support, so I needed the kernel and initrd on the ESP, and I figured a UKI was a convenient way of bundling them. Anyway, yes, if you set the `options` parameter in a type1 entry it will override the cmdline specified in a UKI.

**Comment by TriMoon** [comment]
> Anyway, yes, if you set the `options` parameter in a type1 entry it will override the cmdline specified in a UKI.

IMHO, that should not be possible, but who knows maybe it's just me that thinks so...
